### ISSUE TITLE
Write toml files with \r\n line endings on Windows

### ISSUE THREAD

**Issue by davidanthoff** [JuliaLang/Pkg.jl#934]
I think the ideal situation would be:
1. if the file already exists on disc, and you update it, use the same line ending style that is used in the file already
2. if it is a new file, use the platform default line ending (i.e. \r\n on Windows, \n everywhere else)

The main benefit would be that if these files are under git version control, and one has installed git with the default settings on Windows, then things are easier if these files have \r\n line endings: if I checkout a repo with these files in it, git will create these files with \r\n endings, because they are detected as a text file. If I then do say a `update`, then the file gets rewritten, but now with a \n line ending, even if nothing else changed. Then git shows me the file as modified, but the only change is the line ending, which is kind of annoying.

**Comment by StefanKarpinski** [comment]
Do you have git setup to checkout text files with CRLF line endings?

**Comment by davidanthoff** [comment]
Yes, that is the default you get on Windows. The git setup presents that as the recommended option, so I would expect that most users on Windows run with that setting.

**Comment by davidanthoff** [comment]
Was this fixed?

**Comment by davidanthoff** [comment]
@fredrikekre bump, just curios, was this fixed and therefore closed?

**Comment by fredrikekre** [comment]
Don't know. Closed since there is no TOML code here any more.

**Comment by davidanthoff** [comment]
I guess the first step would be to add a line-ending option to TOML.jl, but then there would still have to be work done here in Pkg.jl that controls what kind of line-ending should be used in a given write? So it seems to me that this issue here should stay open.

**Comment by StefanKarpinski** [comment]
Can't git handle the `\r\n` translation?

**Comment by StefanKarpinski** [comment]
This feature seems kind of fiddly. The `\r\n` line ending convention is kind of a relic at this point. There is one true text file format, especially for source code: UTF-8 with `\n` newlines. Most dev tools support this these days, even on Windows. We can certainly support reading `\r\n` line endings (and we already do), but it seems far better if there is one canonical output format for each TOML file, not different optional variations on it. It seems like we could recommend that people use `core.autocrlf = input` so that everything is checked-out as-is, but if any editors introduce `\r\n` line endings, git will automatically convert them to `\n` before committing.

**Comment by davidanthoff** [comment]
> The `\r\n` line ending convention is kind of a relic at this point.

I think that is really not true on Windows. If you just use git with the default configuration on Windows, all your text files will have `\r\n` line endings. Same is true for VS Code on Windows: you create a new file, it is CRLF by default. I don't think we should encourage all Julia users on Windows to change a global git configuration setting for things to work well, because a) we'll never be able to get that message to everyone, and b) the setting also applies to anything non-Julia, and so there might be lots of scenarios where people just want to stick with the default for non-Julia reasons. I think at the end of the day Julia should just work smoothly with whatever the default is for git on a given platform.

I think the nicest behavior would be something like this:
- if Pkg.jl edits an existing file, it just uses whatever line-ending was previously used in that file. That is how pretty much all text-editing dev tools on Windows work, AFAIK.
- for creating new files, I'm not sure whether it would make more sense to just always use CLRF on Windows, or look at the global git config values and use that.

Yes, that is more involved than one would wish, but I think all text editing tools on Windows just do it that way?

**Comment by StefanKarpinski** [comment]
Ok, if someone wants to figure out how to read the git setting (via libgit2, presumably?), adding an option to TOML to emit `\r\n` line endings shouldn't be hard, but figuring out reading the setting should come first since otherwise we don't really need the TOML option.

**Comment by KristofferC** [comment]
> If you just use git with the default configuration on Windows, all your text files will have \r\n line endings. 

I installed Git on a new Windows machine. The default choice was:

![Screenshot-20200904144802-473x155](https://user-images.githubusercontent.com/1282691/92241318-5544f180-eebe-11ea-8649-7d60693850ff.png)

I then devved a package, did some package operations and undid them. `git status` showed no status change. So from a new installation of Windows, this is apparently not a problem.

**Comment by KristofferC** [comment]
I wrote some stuff here earlier that was a bit wrong. Anyway, I think for existing files just replacating the existing could be done easily. For new files, I think just using `\n` is fine.

**Comment by davidanthoff** [comment]
Could we for now just reopen the issue here?
